canarytokens/channel_dns.py

```diff
@@ -238,6 +240,18 @@ def query(self, query: Query, src_ip: str):  # noqa C901
         ):
             return defer.succeed(self._do_dynamic_response(name=query.name.name))
 
+        if (canarydrop.type == TokenTypes.CMD) and (
+            invocation_id := src_data["src_data"].get("cmd_invocation_id")
```

Comment:
Would this perhaps be equivalent to:
`
invocation_id = src_data["src_data"].get("cmd_invocation_id")
if (canarydrop.type == TokenTypes.CMD) and invocation_id:
`

If so, let's rather do this:
`
 if (canarydrop.type == TokenTypes.CMD) and (
            invocation_id := src_data["src_data"].get("cmd_invocation_id") is not None
):
`

Reply:
They are roughly equivalent (except that we don't check the invocation ID for all token fires, the check only happens if it's `TokenTypes.CMD`). The binding priorities there are a little awkward, since the `:=` applies after the `is not None` (making the `invocation_id` a `bool` instead), but the `is not None` applies after the `and`, so we need an extra set of parens:
```py
        if (canarydrop.type == TokenTypes.CMD) and (
            (invocation_id := src_data["src_data"].get("cmd_invocation_id")) is not None)
```

However, this `invocation_id` is only extracted if there's an 8-character alphanumeric string in the right place in the domain, so if it's not `None` it will never be Falsy...

Reply:
In that case we're doing that `.get()` on almost every DNS request we get... and I'm not sure they'll all have `src_data["src_data"]`

So we can either go with
```python
invocation_id = src_data.get("src_data", {}).get("cmd_invocation_id", None)
if all(( 
    canarydrop.type == TokenTypes.CMD, 
    invocation_id is not None 
)):
```
or
```python
if canarydrop.type == TokenTypes.CMD:
    invocation_id = src_data["src_data"].get("cmd_invocation_id", None)
    if invocation_id is not None:
        ...
```
I lean toward the latter as it's still readable and doesn't do the lookup unless it's needed.